### PR TITLE
Check state of database connection on error and exit if it is unusable

### PR DESCRIPTION
**Pull request by pontus**

### CODE REVIEW

**Review by blankdots — APPROVED** [review]

**Comment by jbygdell** [comment]
```diff --git a/internal/database/db.go b/internal/database/db.go
index 045a073..7fda164 100644
--- a/internal/database/db.go
+++ b/internal/database/db.go
@@ -11,4 +11,5 @@ import (
        "path/filepath"
        "strings"
+       "time"
 
        log "github.com/sirupsen/logrus"
@@ -28,5 +29,6 @@ type Database interface {
 // SQLdb struct that acts as a receiver for the DB update methods
 type SQLdb struct {
-       DB *sql.DB
+       DB       *sql.DB
+       ConnInfo string
 }
 
@@ -76,5 +78,5 @@ func NewDB(config DBConf) (*SQLdb, error) {
        }
 
-       return &SQLdb{DB: db}, nil
+       return &SQLdb{DB: db, ConnInfo: connInfo}, nil
 }
 
@@ -103,4 +105,21 @@ func buildConnInfo(config DBConf) string {
 }
 
+// Ping verifies a connection to the database is still alive,
+// establishing a connection if necessary.
+func (dbs *SQLdb) ping() (*sql.DB, error) {
+       db := dbs.DB
+       if err := db.Ping(); err != nil {
+               db.Close()
+               time.Sleep( 5 * time.Second)
+               log.Debugf("reconnecting to DB with <%s>", dbs.ConnInfo)
+               db, err := sqlOpen("postgres", dbs.ConnInfo)
+               if err != nil {
+                       return nil, err
+               }
+               return db, nil
+       }
+       return db, nil
+}
+
 // GetHeader retrieves the file header
 func (dbs *SQLdb) GetHeader(fileID int) ([]byte, error) {
@@ -218,5 +237,15 @@ func (dbs *SQLdb) MapFilesToDataset(datasetID string, accessionIDs []string) err
        const getID = "SELECT file_id FROM local_ega.archive_files WHERE stable_id = $1"
        const mapping = "INSERT INTO local_ega_ebi.filedataset (file_id, dataset_stable_id) VALUES ($1, $2);"
-       db := dbs.DB
+
+       log.Debugln("ping database")
+       db, err := dbs.ping()
+       if err != nil {
+               log.Errorf("database unreachable, reason: %v", err)
+       }
+       if err := db.Ping(); err != nil {
+               log.Errorf("ping error: %v", err)
+       }
+
+       log.Debugln("start transaction")
        var fileID int64
        transaction, _ := db.Begin()```

File: internal/database/db.go [hunk 1]
```diff
@@ -102,18 +107,48 @@ func buildConnInfo(config DBConf) string {
...
+		time.Sleep(5 * time.Second)
+		log.Debugf("Reconnecting to DB with <%s>", dbs.ConnInfo)
+		dbs.DB, _ = sqlOpen("postgres", dbs.ConnInfo)
+	}
+
+}
+
 // GetHeader retrieves the file header
 func (dbs *SQLdb) GetHeader(fileID int) ([]byte, error) {
 	db := dbs.DB
```
Comment (jbygdell) [inline internal/database/db.go:140]: Missing check.

```suggestion
	dbs.checkAndReconnectIfNeeded()

	db := dbs.DB
```

**Review by jbygdell — CHANGES_REQUESTED** [review]

**Comment by pontus** [comment]
I'm getting rid of checkAndAbortOnBadConn and using wrappers for all calls that retries a number of times before giving up. I'm happy to take opinions on that.

**Comment by jbygdell** [comment]
> I'm thinking of getting rid of checkAndAbortOnBadConn and using wrappers for all calls that retries a number of times before giving up. I'm happy to take opinions on that.

Plan to do this in this PR?

**Review by jbygdell — APPROVED** [review]

**Comment by pontus** [comment]
Open to suggestions, but currently, I'm leaning towards merging this and do a new PR for what we want.

**Comment by blankdots** [comment]
I suggest a new PR so that we can move this forward

**Pull request merged**